assets/src/components/v2/disruption_diagram/disruption_diagram.tsx

```diff
@@ -45,22 +55,603 @@ interface MiddleSlot {
+    fill="#171F26"
+    stroke="#E6E4E1"
+    strokeWidth="3.58209"
+  />
+);
+
+const ShuttleStopIcon: ComponentType<IconProps> = ({ x }) => (
+  <circle cx={x} cy="24" r="10" fill="white" stroke="#171F26" strokeWidth="4" />
+);
+
+const StopIcon: ComponentType<IconProps> = ({ x, className }) => (
+  <circle
+    cx={x}
+    cy="24"
+    r="10"
+    fill="white"
+    className={className}
+    strokeWidth="4"
+  />
+);
+
+const StopIconEndpoint: ComponentType<IconProps> = ({ x, className }) => (
+  <circle
+    cx={x}
+    cy="24"
+    r="20"
+    fill="white"
+    className={className}
+    strokeWidth="8"
+  />
+);
+
+const ArrowEndpoint: ComponentType<IconProps> = ({ x, className }) => (
+  <path
+    transform={`translate(${x} 12)`}
+    width={204}
+    d="M0 24V0H59.446C59.8085 0 60.1642 0.0985159 60.475 0.285014L77.1417 10.285C78.4364 11.0618 78.4364 12.9382 77.1417 13.715L60.475 23.715C60.1642 23.9015 59.8085 24 59.446 24H0Z"
+    className={className}
+  />
+);
+
+const getEndpointLabel = (labelID: string, x: number, isArrow: boolean) => {
+  let labelParts = endLabelIDMap[labelID];
+  if (labelParts.length === 1) {
+    return (
+      <text
+        className="label--endpoint"
+        transform={`translate(${x} -5) rotate(-45)`}
+      >
+        {isArrow && <tspan className="label">to </tspan>}
+        {labelParts[0]}
+      </text>
+    );
+  } else {
+    return (
+      <>
+        <text
+          className="label--endpoint"
+          transform={`translate(${x} -5) rotate(-45)`}
+        >
+          {isArrow && <tspan className="label">to </tspan>}
+          {labelParts[0]}
+        </text>
+        <text
+          className="label--endpoint"
+          transform={`translate(${x + 45} -5) rotate(-45)`}
+        >
+          {labelParts[1]}
+        </text>
+      </>
+    );
+  }
+};
+
+interface EndSlotComponentProps {
+  slot: EndSlot;
+  line: LineColor;
+  isCurrentStop: boolean;
+  isAffected: boolean;
+  effect: Effect;
+}
+
+interface FirstSlotComponentProps extends EndSlotComponentProps {
+  spaceBetween: number;
+}
+
+const FirstSlotComponent: ComponentType<FirstSlotComponentProps> = ({
```

Comment:
FirstSlotComponent & LastSlotComponent are almost identical, except for that FirstSlot includes a background. I think they could be generalized into EndSlotComponent.

Honestly, MiddleSlotComponent is also very close and maybe could be incorporated, too, but I could be overlooking something